Enhance add Gui Translations

SlySven opened the pull request:
Right this steaming pile of _code_ represents my reworking of https://github.com/SlySven/Mudlet/tree/enhance_refactorForI18n (that was left to fester in a dark place a couple of years ago) and this should go pretty far towards the framework needed to use languages other than English (American) for the User Interface, it includes partial translation files (both source `./src/mudlet_`xx_YY`.ts` and binary `./src/mudlet_`xx_YY`.qm`) for:
+ en_US (American English)
+ en_GB (British English)
+ fr_FR (France French)
+ de_DE (German German)
+ ru_RU (Russian Russian)
+ zn_CN (Simplified Chinese)

![screenshot_20171012_045042](https://user-images.githubusercontent.com/6163092/31478542-fc17f078-af08-11e7-92de-5c67b6312f39.png)

At present the `.qm` files need to _appear_ to be in the same directory as the mudlet executable for Linux and Windows builds - I left some macOs defines in place which will not help.  Currently, if no suitable files are found on start-up the control to switch on translation/change language will be disabled:
![screenshot_20171012_051809](https://user-images.githubusercontent.com/6163092/31479072-c677d9d4-af0c-11e7-8a94-a7d2581c84b8.png)

During development I have found it is enough to set up some symbolic links to the `./src/mudlet_xx_YY.qm` files from the "build" directory where the mudlet executable is built then you will get:
![screenshot_20171012_052251](https://user-images.githubusercontent.com/6163092/31479186-718a21ce-af0d-11e7-8ecd-c7d345d5bd54.png)

For demonstration purposes the most obvious (and "complete"; although that complete is only in terms of the %-age of entries - the translation quality is probably 💩 !)  language to try is French.  Note that that control is filled with entries in the language of the particular language itself as nothing else makes sense from a UI point of view; as a help for this the tool-tip for that control is a special one in that it is NOT translated but instead contains a series of texts all saying the same thing but in a range of languages...
![screenshot_20171012_053317](https://user-images.githubusercontent.com/6163092/31479436-237fa95c-af0f-11e7-9b2b-6e44857b6617.png)

There are a couple of _External_ commands that are included in the Qt Creator IDE by default that seem to be to do with:
- extracting the `QString`s that need to be translated from various locations in the source code {`lupdate` - produces/updates `.ts` files which are XML files} 
- convert the translation data in `.ts` files into the binary `.qm` files that are needed by the application at runtime to provide the data needed in a particular language

however as provided the first one reads the `TRANSLATIONS` variable in the `qmake` project file and processes all the languages to produce what are _full_ translation files with every `QString` in - for the `English (American)` language, as the source code is already _in_ that language it produces a much larger file than is needed, and by running the `lupdate` command with an existing `mudlet_en_US.ts` file it can be made much simpler with the additional `-pluralonly` option that _only_ picks out the cases where the `%n` parameter is used to specify a (usually integer but it need not be) __3rd__ argument to the `QObject::tr()` method which enables the handling of all the different _plural_ forms for those ___many___  languages where adding an `s` if there is more (or less) than __one__ article is just plain wrong!

also, the second command has a useful option - that is not used by the default configuration in Qt Creator - that will mark a case where there is not a "finished" translation in place for the particular `QString` instance, which, when debugging a translation, or where the source code string has been modified (which invalidates the translations based on it) can be prefixed in the application to flag it as something that needs attention.

Anyhow, I have included in this PR a couple of simple *nix  shell scripts which can be added to the Qt Creator "External" tool list, the first one is `unix_lupdate.sh` and you can set it up like this:
![screenshot_20171014_041748](https://user-images.githubusercontent.com/6163092/31572148-fd6ddfc0-b096-11e7-8e47-e66c6e222e85.png)

It takes one argument - the location of the current Qt library version's `bin` directory so it knows which version of `lupdate` to run - and it will cycle through all the `mudlet_`xx`.ts` and `mudlet_xx_YY.ts` files, updating them (whether mentioned in the qmake project file or NOT) in the mudlet `/src` directory and pick out the `mudlet_en_US.ts` for the special _plurals only_ treatment.

The other script `unix_lrelease.sh` also takes the same path to wanted binaries argument and if run like that it will regenerate all the `mudlet_`xx`.qm` and `mudlet_xx_YY.qm` binaries but by default it will include a `#` prefix on all the unfinished translations - as does `QLinguist` when it is previewing a form/dialogue.  However you can change this with an optional second argument, I have been using `~` to use a different marker, to suppress any prefix supply an empty string `""` as a second argument.

The work flow when C++ coding is:
+ write/modify source which changes a `QObject::tr(`...`)` or other translatable `QString`
+ run `lupdate` or my `unix_lupdate.sh` script - by default they will not "throw away" obsolete translations i.e. ones where the source string in the `tr(`...`)` command is changed so it will not dump already worked out translations (there is a command line option to do so, but it is an all-or-nothing action though it is possible to hand edit EVERY `.ts` file if it is vital that a really wrong (accidentally obscene source text word :devil:) thing has to be purged, but that is not a thing to be undertaken lightly!
+ run `QtLinguist` and enter a translation for the languages you are invested in/investigating, if happy with it, mark the translation as "finished" (will have a green or yellow tick)
+ run `lrelease` or my `unix_lrelease.sh` script
+ run `mudlet` and select the language (if it is not already the one that was stored as being set)

__We will probably have to discipline ourselves in making commits to the `.qm` files stored in the repository - it may be useful to hash out a system where developers can keep a local "working" copy of a set of `.ts`/`.qm` files if they are working on GUI stuff that will hit translated `QStrings`__

In passing I would note that I have addressed the "other licences" issue #1273 herein.

This is a big one - please ask questions if something doesn't make sense - I regret that the main commit message is a bit of a mess as the code evolved over time and I lost some stuff and had to retrieve it from the equivalent to the git _rubbish bin_.

### ToDo list:
+ [ ] - insert "debugger option" to specify/override where to look for mudlet_xx_YY.qm files at runtime {and improve handling of different Os wanting them in different places at the same time; needed for macOs based developers...}
+ [ ] - do a working 3.4+ version of #436 and merge it in , as having reformatted the help messages for (completely) empty types items in the editor the fixed size widget at the top of the screen is no longer big enough to display the, now arranged into an HTML list, set of instructions for some types.
+ [ ] - add a command line argument to set/change the language from the command line, to provide a get-out if the user really cannot navigate to the setting from the start (or after installation) and they have fouled it up and need to adjust the setting.

vadi2 commented:
Hi Stephen! Thanks for doing this massive amount of effort, it is awesome to see thie PR finally come to life :)

What do you think about adding the translation files into the Qt Resource file - that's how we've already been solving the runtime detection issue for files already.

vadi2 commented:
Let's leave #436 for later - it addresses a totally different problem and has many issues itself that'll slow down this PR from making it in. If we do this step by step, it'll be easier.

SlySven commented:
On blast got an `ENOCOFFEE` with a couple of:

    #if Q_OS_WIN32

instead of:

    #if defined(Q_OS_WIN32)

causing fails on Apveyor and it turns out I use `zip_file_add(`...`)` and `zip_dir_add(`...`)` in an included fix for the buggy module/package exporter that silently breaks if there is a sub-directory in the package being made but those are libzip 1.x functions that replace the declared obsolete 0.1x `zip_add(`...`)` ones but get which version is used on `trusty` on Travis... 😒 🔨

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -4319,25 +4405,50 @@ void T2DMap::slot_setCustomLine()
+                                       .arg(tr("<p>The room this special exit leads <b>to</b>.</p>")));
+    pTreeWidgetItem_header->setToolTip(2, QStringLiteral("<html><head/><body>%1</body></html>")
+                                       .arg(tr("<p>The command or Lua script that goes to the given room.</p>"
+                                               "<p>It is what Mudlet will use to speedwalk (route find) using this particular exit so it can be a complex script - however <i>currently</i> Mudlet also uses this entry to identify this particular exit so if it is to be a <i>scripted</i> procedure to use this exit, a Lua function to do so is highly advisable!</p>")));
```
I don't think many people will understand what this means.

SlySven replied:
Currently we do not provide a means to have a name for a special exit that is an alias for a script - I recall that there is a comment somewhere in the C++ code for special exits that _"...a script should be entered with a `script: ` (or was it a `lua:`) prefix..."_ but that is confusing and wrong in that __some__ lua packages use that convention but it is not something that is hard coded in the C++ core.  It also means that if the commands to use a special exit are something like `unlock trapdoor; climb down; lock trapdoor` then THAT is the identifier that has to be used/displayed in the GUI associated with the mapper - however I would like to see that changed so that there can be an identifier `trapdoor` that can be used for display purposes associated with the original `command`...

... so at some future stage I hope to be able to change this text! 😀

vadi2 replied:
How about the following text:

> Mudlet will use this command in walking through this exit: it can be something simple as a direction or something more capable like a Lua script.

SlySven replied:
☑️  Sentence emended to: "Mudlet will use this command in walking through this exit (when requested to find a route through the map); it can be something simple as a direction or something more capable like a Lua script."

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -605,7 +601,43 @@ const QMap<QString, QPair<QString, QVector<QChar>>> TBuffer::csmEncodingTable =
+};
+
+QMap<QString, QString> TBuffer::smEncodingNamesMap  = {
+    {QStringLiteral("ASCII"), tr("ASCII (Telnet default encoding)", "Server encoding name map: Ensure both instances have same translation text (1 of 2)")},
```
I don't know what do you mean here.

SlySven replied:
The `Developer's comment` ? I have found it is advisable to have something like this to ensure that BOTH occurrences of `ASCII (Telnet default encoding)` get modified in step - if this instance of this source string gets modified it highlights that there is an instance in this or another class that ALSO must be both edited and translated as well.  By default if there was not such a comment then only the first case of `ASCII (Telnet default encoding)` gets displayed in `QtLinguist` by default and if ONLY this instance in the source code gets edited then there can be programme breakage when the other instance gets used.  This is the issue when using `QStrings` as identifiers internally - and why `QFlags`/`enums` are a better solution (c.f. custom line styles...)

vadi2 replied:
Ah ok, how about saying "ensure this text is translated the same"?

SlySven replied:
- need to keep the X of Y part but in principle I guess I can shorten it - and invalidate all the current translations now using it... 🙁

vadi2 replied:
Sorry, but these kinds of adjustments would happen given they were made before the PR was even up. Isn't it possible to just find & replace in a file?

SlySven replied:
Well there is a multiplication factor i.e. change anything in a `tr(`...`)` and it is not just that source file but every (well apart from the _plurals only_ `mudlet_en_US.ts`) file changes as well - before I pushed out the PR I got to do it all behind the scenes; however going forward coders ought to at be considering _"do I really need to change this QString?"_

TBH I suspect we will have to re-examine a lot of the textual content in Mudlet and once this PR goes in I think a couple of multi-lingual Mudlet developers will need to go through and do a translation in, say German and Russian 😉 just so we can identify stuff that is not flagged for translation but which should be.  This will likely require us to rethink some code fragments (other languages strings will not necessary be the same size on screen which may break some dialogs) and I suspect it will better all around if we do this for a couple of languages before we open it up for collaborative Web translation so we are not asking people to work on strings that have to be thrown away...

 At some stage we will want to include the defines for `QT_NO_CAST_FROM_ASCII` and `QT_NO_CAST_TO_ASCII` in the CMake/QMake project files but we can do that on a class by class basis at the moment.

vadi2 replied:
I don't understand why can't we revise the translation hint text to make sense to translators.

You have to understand not everybody is a technical coder who understands what terms like "instance" will mean.

SlySven replied:
Fair enough - I'll give it a tweak...

vadi2 replied:
Thank you 👼

SlySven replied:
☑️ Have changed all _instance_ to _case_ so that there is a fairly standardised: 

> This text is used programmatically, ensure all cases have same translation (x of y)

normally this is to signal that the C++ code relies on the same (translatable) text to be used at different points in the codebase and should only be encountered in dialogues which are only around for a short time and it is not worthwhile to try and code to handle the text being changed by a GUI language change happening between where one ~instance~case of the string is used and another.

In this case _that_ is not the case, instead it is because the use is in two separate classes (`TBuffer` and `dlgProfilePreferences`) and it may not be obvious that they are linked and a developer changing one ~instance~case needs to remember to change the other as well; so the text is now:

> Server encoding name map: ensure all cases have same translation text (x of y)

😀

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -779,8 +811,8 @@ TBuffer::TBuffer(Host* pH)
-    for (auto table : csmEncodingTable) {
-        Q_ASSERT_X(table.second.size() == 128, "TBuffer", "Mis-sized encoding look-up table.");
+    for (auto table : csmEncodingsMap) {
+        Q_ASSERT_X(table.size() == 128, "TBuffer", "Mis-sized encoding look-up table.");
```
We should change this to the compile-time `static_assert()` available in C++11: faster feedback.

SlySven replied:
Does it not get MACROed out to that _when it can be_ by the Qt Libaries?

vadi2 replied:
What do you mean by when it can be - we're already using C++11, there's no way it can't be.

I don't think it is, Qt documentation says nothing about it. I think this is a runtime check.

SlySven replied:
I see what you mean with the difference between a compile and run-time checks and yes it would be better to check it once during compilation than every time a `TBuffer` is constructed (especially as the data is a globally shared single instance).

I am just not quite sure of the best way to reliable ensure that each table in the set is exactly 128 elements because if it is not then a source code edit has messed something up and it will cause _undefined behaviour_ if the table is shorter and probably data corruption if it is longer...

vadi2 replied:
I'm not sure either, but this would be a learning opportunity. I think a combination of a constexpr function + static_assert would do it.

vadi2 replied:
I've had a look, I think we need C++14 at minimum to do this. Either way, this is not something you have to address for this PR for clarity's sake.

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -3734,8 +3766,7 @@ QString TBuffer::bufferToHtml(QPoint P1, QPoint P2, bool allowedTimestamps, int
         // formatting according to TTextEdit.cpp: if( i2 < timeOffset )
-        s.append(R"(<span style="color: rgb(200,150,0); background: rgb(22,22,22); )");
-        s.append(R"(font-weight: normal; font-style: normal; text-decoration: normal">)");
+        s.append("<span style=\"color: rgb(200,150,0); background: rgb(22,22,22); font-weight: normal; font-style: normal; text-decoration: normal\">");
```
We're not using `tr()` here - why was the raw literal removed?

SlySven replied:
I have a nasty suspicion that the raw literals are actually buggering up the parsing of the file for `tr(`...`)` instances **even when they are not involved in the using of those literals**.  As yet I have not found any third party utility that gets around this Qt Bug...

vadi2 commented on src/TBuffer.h:
```diff
@@ -149,9 +161,12 @@ class TBuffer
-    static const QString& getComputerEncoding(const QString& encoding);
+    static const QList<QString> getComputerEncodingNames() { return csmEncodingsMap.keys(); }
+    // Found issues with returning a const QString reference from the global
+    // static table - so keep it simple and reliable and return by value which
```
Read up on C++11 move semantics ([example](https://mbevin.wordpress.com/2012/11/20/move-semantics/)) as well.

SlySven replied:
But this is not a move but a copy - we do not want to destroy the original... 😖

vadi2 commented on src/TConsole.cpp:
```diff
@@ -768,7 +816,7 @@ void TConsole::closeEvent(QCloseEvent* event)
 
             if (std::get<0>(result) == false) {
-                QMessageBox::critical(this, tr("Couldn't save profile"), tr("Sorry, couldn't save your profile - got the following error: %1").arg(std::get<2>(result)));
+                QMessageBox::critical(this, tr("Could not save profile"), tr("Sorry, could not save your profile - the following error was found:\n%1.").arg(std::get<2>(result)));
```
Why the change?

SlySven replied:
Ease of comprehension by (machine) translators IMVHO - and as it is an undesired occurrence "Sorry mate I've destroyed your game data 'cos: xxxxx" doesn't carry quite the same seriousness of "Sorry, it has not been possible to save that because: xxxx" - I cannot recall where but I was always taught to avoid trivial contractions in program output (using `'` only saves one space but it does degrade readability scores as far as I understand it) and "got the following error:" is, I suspect, harder to translate than the better formed "the following error was found:" IMHO .

vadi2 commented on src/TConsole.h:
```diff
@@ -254,7 +254,9 @@ class TConsole : public QWidget
     TSplitter* splitter;
     bool mIsPromptLine;
-    QToolButton* logButton;
+    QToolButton* mTimeStampButton; // was an auto (local to constructor) pointer timeStampButton
```
Don't need to know what it was before. `git blame` can help us with that.

SlySven replied:
When I am coding and changing a member name around so that I can remember what it was called at the start of my ~~hacking~~coding session - but yeah I can see how it leads to historical cruft...

vadi2 replied:
:+1:

SlySven replied:
☑️ Have cleaned out a number of `// Was` type entries I had left behind - but I ought to document variable scope/renames in the commit messages as I find it useful when trying to merge/rebase/cherry-pick to remember what the identifier was at a particular point in time... 😌

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -10320,6 +10347,19 @@ int TLuaInterpreter::setDefaultAreaVisible(lua_State* L)
+    const QString languageCode(mudlet::self()->getGuiLanguage());
+
+    if (languageCode.isEmpty()) {
+        lua_pushstring(L, "none");
```
What about `nil` instead?

SlySven replied:
💭 Well that would be _one_ way to do it - but then a script would have to respond to the _nil_ specially - OTOH it might be worth it so that _nil_ means that there ARE no translations available as opposed to "none" meaning it has been "switched-off"...

vadi2 replied:
I don't see how that would matter to an API. Let's go with `nil`, if we get requests specifically for "none" we can add it.

SlySven replied:
I have only give a getter for the language (because it is common across all active profiles and it is not something that a profile should be able to change, only to detect) but if there were to be a setter as well it might be worth returning a `nil` when there are __no__ translations available but an English as per Lua API requirements "none" when there are but they have been disabled... does that seem sensible even thought I am not planning a setter for GUI Language?

vadi2 replied:
Actually neither `nil` nor `none` make sense. You can't have "no language". It'll be English at minimum...

SlySven replied:
Well, it depends on whether you always want an en_US translation (which is plurals only) or the source code text (if you are debugging a translation) - the difference is when the `%n' replacement parameter is used in `tr(`...`)` see the "Handling Plurals" section in [I18n source translation](http://doc.qt.io/qt-5/i18n-source-translation.html)

In both the "no translations available" or the "none selected case" uses will see messages like:

    0 Room(s) selected
    1 Room(s) selected
    2 Room(s) selected

but if the en_US translation is available this should change to the cleaner/clearer:

    0 Rooms selected
    1 Room selected
    2 Rooms selected

whilst the end user may never want to run with out a translation loaded, we developers likely _will_ want the ability to turn off the translation infrastructure...!  From a diagnostic point of view a package creator who is contemplating supporting multiple languages they may want to know the difference between knowing that the user does have but is not using a particular translation compared to a possibly mis-configured setup where the executable cannot find the auxiallary translation files.

:bulb: As well as providing the current translation code (and the previous one in the Mudlet change event) perhaps we should also provide a list of the codes available...?

vadi2 replied:
As we have always had, there always has been a language available - so `en_US` will always be at least available. Thus there should never be `nil` or `none` returned I think is the way to go.

If a developer wants to debug this, they can fiddle around with it as needed. If we find we need to do it often, we'll put in tools - but right now we have no idea and let's err on the side of less work, less maintenance, less issues.

SlySven replied:
> As we have always had, there always has been a language available

I am trying to say that is not exactly the case - we have always had the texts included in the source code and that is English-ish and largely to the American version - however we will be including a translation explicitly for that locale which only differs in the handling of plurals (to make them look nicer) but the latter is only going to be available provided the translations are found when __mudlet__ starts up, should something fail in that process (partial install, permissions problems, etc.) the executable will fall-back to running without, (admittedly that is not likely to be an issue for those users/platforms that accept our likely action to include the translations in the resources)...

vadi2 replied:
Yeah but even in that case, Mudlet is will be using some language, not `none` or `nil` - so we can assume it to be `en_US`, even if it's not technically 100% correct. Nobody will mind, and it'll reduce the confusion we'd get from seeing a `none` language

vadi2 replied:
This still needs to change - "none" is not a language people speak! Let's make this for humans, not engineers :wink:

vadi2 commented on src/TRoom.cpp:
```diff
@@ -35,6 +35,33 @@
+{
+    quint8 temporary;
+    ds >> temporary;
+    switch(temporary) {
```
Wouldn't it be a lot simplier to initialise it to `Qt::SolidLine` and have an `if` for `Qt::DashDotDotLine`?

SlySven replied:
Allows a fall-through to a meaningful/usable value should something else show up (there are a couple of other `Qt::PenStyles` {none and custom} which we cannot usefully use...

vadi2 replied:
Okay, but see my comment - just an initialisation and an _if_ will take a lot less space, and it won't give warnings on GCC.

SlySven replied:
Seems like there ought to be a Qt provided MACRO for this that can drop in an appropriate `[[`..`fallthrough]]` "attribute" which GCC7 is to get alongside Clang which has had it's own for a while now...

vadi2 replied:
I believe that got added to C++17, but we unfortunately can't use it yet because Linux distros are keeping us down. That said we can fine our own fallthrough macro right now that'll work: https://infektor.net/posts/2017-01-19-using-cpp17-attributes-today.html#using-the-fallthrough-attribute

SlySven replied:
Noted, though where would we define it? It is hard to code a `#define` macro that takes _arguments_ in the QMake `mudlet.pro` file IIRC...

vadi2 replied:
The point I'm making here is that we don't need this complicated switch :smile: we can simply do:

```
value = Qt::SolidLine;
if (Qt::DashDotDotLine) {
   value = static_cast<Qt::PenStyle>(temporary);
}
```

3 lines of code compared to 12

SlySven replied:
That code doesn't make sense - the `switch` is being used to validate 4 specific values of `temporary` {the, cast to a `quint8`, values of `Qt::DotLine`, `Qt::DashLine`, `Qt::DashDotLine`, `Qt::DashDotDotLine`} and to use which ever of __them__ if they are found plus a fifth value which is used __for all other (252) cases__ (including the remaining valid one of `Qt::SolidLine`)...

Going for an `if` based it would be:

    if ( static_cast<Qt::PenStyle>(temporary) == Qt::DotLine || static_cast<Qt::PenStyle>(temporary) == Qt::DashLine || static_cast<Qt::PenStyle>(temporary) == Qt::DashDotLine || static_cast<Qt::PenStyle>(temporary) == Qt::DashDotDotLine ) {
        value = static_cast<Qt::PenStyle>(temporary);
    } else {
        value = Qt::SolidLine;
    }

In this situation I actually find the `switch` one clearer. 😲

(It is only the need to tell suspicious compliers that the running together of those `case`s is deliberate and wanted which requires three of those lines - I could swap the `[[clang::fallthrough]]` for `; // Fall-through` though and that should shut each one of them up!)

vadi2 replied:
OK, how about the following?

```c++
    switch (temporary) {
    case Qt::DotLine | Qt::DashLine | Qt::DashDotLine:
        [[clang::fallthrough]];
    case Qt::DashDotDotLine:
        value = static_cast<Qt::PenStyle>(temporary);
        break;
    default:
        // Force anything else to be a solidline
        value = Qt::SolidLine;
    }
```

SlySven replied:
Is that a new-fangled C++11 thing, being able to OR `case` value together? Unfortunately I do not think the individual values are 2^n values so they can't be used in the manner of a bitmask either...

vadi2 replied:
No, it's something that was around for a while: https://doc.qt.io/qt-5/qflags.html#operator-7c-eq

SlySven replied:
Looking at the current definition of that enum:
> Constant | Value | Description
> -- | -- | --
> Qt::NoPen | 0 | no line at all. For example, QPainter::drawRect() fills but does not draw any boundary line.
> Qt::SolidLine | 1 | A plain line.
> Qt::DashLine | 2 | Dashes separated by a few pixels.
> Qt::DotLine | 3 | Dots separated by a few pixels.
> Qt::DashDotLine | 4 | Alternate dots and dashes.
> Qt::DashDotDotLine | 5 | One dash, two dots, one dash, two dots.
> Qt::CustomDashLine | 6 | A custom pattern defined using QPainterPathStroker::setDashPattern().

A bitwise-OR doesn't work to provide a `case` value that can be used to detect `Qt::DotLine` OR `Qt::DashLine` OR `Qt::DashDotLine` unless C++11 has gained something that I am not aware of..

vadi2 replied:
Oh. Nevermind me then!

vadi2 commented on src/TTextEdit.h:
```diff
@@ -85,11 +85,11 @@ class TTextEdit : public QWidget
     void setConsoleBgColor(int r, int g, int b) { mBgColor = QColor(r, g, b); }
     void setIsMiniConsole() { mIsMiniConsole = true; }
-    void copySelectionToClipboardHTML();
+    // Returns whether the timestamps are showing AS A RESULT OF the call:
```
What do you mean by 'as a result of the call'?

SlySven replied:
This class needs to know what the time-stamp visibility setting is and it is private IIRC so passing the value back says having to change the visibility _just_ for this.

vadi2 replied:
`AS A RESULT OF the call` sounds like the call will have side-effects is what the comment sounds like to me. I think just `// Returns whether the timestamps are showing` will work, or if the function toggles _and_ returns the value, `// Toggles timestamps on/off and returns the current state``

SlySven replied:
`// Toggles timestamps on/off and returns the current state` works for me. 😁

SlySven replied:
☑️ Done

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -1265,88 +1304,108 @@ void cTelnet::postMessage(QString msg)
+            // stored in mPrefixLength but we still check for the other
+            // components to confimrm validity:
+//            quint8 prefixLength = hyphenIndex + 1;
+//            while (body.at(0).at(prefixLength) == ' ') {
```
Still need this?

vadi2 commented on src/dlgAboutDialog.cpp:
```diff
@@ -94,4 +104,314 @@ dlgAboutDialog::dlgAboutDialog(QWidget* parent) : QDialog(parent)
+    // I have found the hard way that the lupdate utility to extract translatable
+    // strings from the code does not like C++11 syle raw string literals, see:
+    // https://bugreports.qt.io/browse/QTBUG-42736
+    QString aboutMudletHeader(tr("<tr><td><span style=\"color:#bc8942;\"><b>Homepage</b></span></td><td><a href=\"http://www.mudlet.org/\">www.mudlet.org</a></td></tr>\n"
```
How come this was moved here? This significantly obstructs the ability to edit it.

SlySven replied:
Have you seen the HTML that this would get converted to by the QtDesigner plugin 😱 - and a translator - whether using QtLinguist __or an external system__ would get __that__ to work on, it does compromise the ease of editing but it reduces the pain at translation time...

vadi2 commented on src/dlgIRC.cpp:
```diff
@@ -196,7 +203,7 @@ QPair<bool, QString> dlgIRC::sendMsg(const QString& target, const QString& messa
 {
     QString msg = tr("Restarting IRC Client");
-    ircBrowser->append(IrcMessageFormatter::formatMessage("! %1.").arg(msg));
+    ircBrowser->append(IrcMessageFormatter::formatMessage(tr("! %1.", "The leading ! is a special status marker used programmatically - not an actual exclamination mark - leave unchanged!").arg(msg)));
```
You're saying the Spanish shouldn't change it to ¡ ? :wink:

SlySven replied:
¡Sí!

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -19,12 +19,30 @@
+ * This should really be called "dlgModuleExporter" - "Packages" should be
+ * reserved for a single node (plus sub-nodes) from one type of Mudlet item that
+ * is "Exported" from the "Editor" whereas "Modules" are collections of possibly
+ * more than one type of Mudlet item and which can include additional reasources
```
That is not correct - packages are not reserved for single-node items, they can be multi-node packages just like modules.

SlySven replied:
How, the "Export" option in the Editor only allows one item to be selected IIRC...

vadi2 replied:
It's a crappy export option...

SlySven replied:
So? That is how it is isn't it, and it does have its uses to get a straight forward XML file. I guess this might be a bit of a simplistic division between the two terms but actually there is some merit in distinguishing between a _simple_ XML "extract of one leaf/branch of one tree" in the Editor compared to a _more complex_ "collection of multiple branches from multiple trees together with other file based resources all zipped up in an archive file"...

vadi2 replied:
I disagree but this isn't the right place to have this discussion, nor is this the i18n PR the right place to sneak remarks into the source code about completely unrelated things :wink:

SlySven replied:
☑️ Reworded to note the difference between what this class does/produces and the "Export" option in the Editor (`dlgTriggerEditor` class) - the latter only exports one "_branch_" of one "_tree_" (widget) of Mudlet items in an `.xml` file whereas this manages a whole "_forest_" (and external files besides) as a `.mpackage` (renamed `.zip`) archive file; but there is no mention of anything about packages vs. modules... 😉

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -262,54 +461,193 @@ void dlgPackageExporter::slot_export_package()
+                    }
+                    fileEntries.remove(QStringLiteral("config.lua"));
+                } else {
+                    qWarning() << "dlgPackageExporter::slot_exportPackage() - ERROR: failed to add package detail file config.lua to archive, the package/module may not work!";
```
It'll work just fine without it, the name will just be assumed from the zip file instead.

SlySven replied:
... but it is a warning that the `config.lua` file has been clobbered...

vadi2 replied:
Yeah, but the text is incorrect.

SlySven replied:
Would:

    qWarning() << "dlgPackageExporter::slot_exportPackage() - WARNING: failed to add package detail file config.lua to archive!"

be acceptable?

vadi2 replied:
:+1:

SlySven replied:
☑️ Done

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -21,11 +21,17 @@
  ***************************************************************************/
 
+/*
+ * THIS FILE CONTAINS UTF-8 UNICODE ENCODED CHARACTER STRINGS THAT ARE (OR
```
I'm not sure what is the takeaway message I'm supposed to get here, could you explain?

SlySven replied:
The tool-tips in the constructor MUST remain as `QStringLiterals(`...`)` and NOT be converted to `tr(`...`)` but I could word it as that now - when I started on this I was not quite sure how it would all work out... 🙂

vadi2 replied:
Okay - can we just say that, without the shouting?

SlySven replied:
~~YES~~ err,  yes!

SlySven replied:
☑️ Done

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -63,6 +69,105 @@ dlgProfilePreferences::dlgProfilePreferences(QWidget* pF, Host* pH) : QDialog(pF
+    QStringList tooltipLanguageEntries;
+
+    // This covers the default en_US as well as en_GB:
+    tooltipLanguageEntries.append(QStringLiteral("Choose the language for Mudlet to use..."));
```
Should be `Choose the language for Mudlet to use:`

elipsis is used to indicate that there is more following in another menu in UI design.

SlySven replied:
Well, I meant the [ellipsis](https://en.wikipedia.org/wiki/Ellipsis#On_the_Internet_and_in_text_messaging) to mean if you select this control it carries out this action`...`

Grammatically I might not be entirely correct, but I suspect that a colon `:` is definitely not right here.

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -325,47 +433,88 @@ dlgProfilePreferences::dlgProfilePreferences(QWidget* pF, Host* pH) : QDialog(pF
+        // Handle case when no map is present:
+        if (pHost->mpMap && pHost->mpMap->mpRoomDB && pHost->mpMap->mpRoomDB->size() > 0 ) {
+            label_mapFileSaveFormatVersion->setEnabled(true);
+            Q_ASSERT_X((pHost->mpMap->mMaxVersion >= pHost->mpMap->mDefaultVersion
```
I think a `static_assert()` would work well here

SlySven replied:
See previous comment about the same Qt Macro elsewhere

SlySven commented on src/KeyUnit.cpp:
```diff
@@ -400,322 +439,347 @@ void KeyUnit::doCleanup()
+    mKeys[0x0f7] = tr("division");
+    mKeys[0x0ff] = tr("ydiaeresis");
+
+    mKeys[0x01000000] = tr("Escape");
```
Not necessarily - a French Key board is not going to have a "BackSpace" key and the "Shift" keys are not labelled "Shift" e.g. see Wikipedia's [AZERTY](https://en.wikipedia.org/wiki/AZERTY) Keyboard layout article...

vadi2 replied:
Okay

SlySven commented on src/T2DMap.cpp:
```diff
@@ -610,8 +614,14 @@ void T2DMap::paintEvent(QPaintEvent* e)
-
+        p.save();
+        p.setPen(Qt::yellow);
+        p.fillRect(0, 0, width(), height(), Qt::darkRed);
```
*sigh* yeah it is a little off-topic - but as I was marking the "No map..." message I remembered being bugged by the layout and colour of the message in the past - and as there are no guarantees about the size (or shape) of the translated message it seemed sensible to use the layout system to centre it properly and allow it to be wrapped if the map was very small rather than specify a _start_ point of half-way across the screen {neglecting the fact the message __has__ a length}...

SlySven commented on src/T2DMap.cpp:
```diff
@@ -1990,40 +2009,40 @@ void T2DMap::createLabel(QRectF labelRect)
 
     QMessageBox msgBox;
-    msgBox.setText("Text label or image label?");
+    msgBox.setText(tr("Text or Image label?"));
```
Well it helps to pick out that there are two options and what those are - and linguistically it might help in translation - imaging you are Zaphod Beeblebrox in Zarniwoop's office: do you leave via the __Door__ or jump out the __Window__?

vadi2 replied:
Just door or window... these aren't names!

SlySven replied:
☑️  Actually have changed it to "Type of label?" - which leads on nicely to the texts on the Buttons which are "Text Label" and "Image Label" as the answers to the question!

vadi2 commented:
I've resolved the merge conflicts that occured, but the PR still has the initial outstanding issues it had in compiling!

vadi2 commented:
Could we move the translations file to a separate top-level `/translations/` to keep things clean?

SlySven commented:
> Could we move the translations file to a separate top-level /translations/ to keep things clean?

In the source tree for development (followed by installation placement in, say, a shared, read only location such as  `/usr/[local/]share/mudlet/translations` for *nix platform builds) ?  That seems like a good idea - :bulb: in conjunction with a "Preferences" => "Special Options" `QLineEdit` to override it for development/testing work .

SlySven commented:
🎉 Finally - it passed CI - ye-ha!

vadi2 commented:
Awesome work!

vadi2 commented:
> What do you think about adding the translation files into the Qt Resource file - that's how we've already been solving the runtime detection issue for files already.

SlySven commented:
That has some promise - though I think we'd want to enable the `-compress` option to `lrelease` as uncompressed I think I was seeing files of size of a couple of hundreds of kilobytes for the binary translation files as they were...

Note that some platforms would not want that though, Linux distributions such as Debian or Ubuntu  expect those shareable, read-only files to be located elsewhere in the file-system and I expect @csmall would patch the source to strip them out of the mudlet binary, probably to `/usr/share/mudlet/translations/` and trim the `mudlet_` off the name...

A `QLineEdit` on the special options does seem to be the way to go on this - with a `:/transations/` default value... which packagers can patch to say `/usr/share/mudlet/translations/` instead but still allow modifications for developers/translators to test locally generated ones in a different (or even `./src/` location).

vadi2 commented:
Who are we gonna share Mudlet translations with exactly?

vadi2 commented:
To make it easier to see which comments of mine are actually requesting changes, I've went through and made a list - hope it helps:

https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686207
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686404
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686454
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686518
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686669
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686689
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686803
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686840
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144686877
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144687001
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144687073
https://github.com/Mudlet/Mudlet/pull/1334#discussion_r144687117

SlySven commented:
> Who are we gonna share Mudlet translations with exactly?

Do you mean, why place them in a `/usr/share/` or `/usr/local/share/` directory? If so it means other users of the same *nix PC who also want to play MUDs - though this is not usually simultaneously with [multi-seat](https://en.wikipedia.org/wiki/Multiseat_configuration) configurations this is becoming increasingly do-able.  The same also applies I suppose in an "install for all users" Windoze installation, though with a centralised `Program Files` location it is likely that the translations can be kept in the same section of the file-system anyhow.

vadi2 commented:
I think it'll be a lot easier to just embed them into a resource file. If someone asks us to split it out and gives pretty good reasoning, we can do that... but otherwise this is additional file loading complication that I'd rather not deal with. This complicates setup instructions, makes it a lot harder for Windows, adds more room for error and frustration, all for no real benefit to us. It seems like a nice idea but really the costs of it are high.

vadi2 commented:
Hey @SlySven, how're you going with this? Did I submit too many comments :(

SlySven commented:
No, (probably) all good feedback - just looking into the ~~abyss~~module handling fixes that got dragged into it (with all the messages/dialogs that needed to be made translatable this poked significantly into the `dlgPackageExporter` class and I addressed some issues there {the complete lack of error handling texts} _as I passed through_ but there was still some things that did not look right which I am delving into at the moment...

SlySven commented on src/KeyUnit.cpp:
```diff
@@ -400,322 +439,347 @@ void KeyUnit::doCleanup()
-    mKeys[0x01020006] = QString("Zoom");
-    mKeys[0x01020001] = QString("Cancel");
+    // This WAS called setupKeyNames but it actually represents all that needs
+    // doing for this class when the language changes...! 8-)
```
☑️  This comment removed.

vadi2 commented:
Nice, thanks for addressing the feedback! I'm also keen on https://github.com/Mudlet/Mudlet/pull/1334#issuecomment-337448117.

SlySven commented:
⏱  I'm looking at the outstanding thing about using a resource file...

vadi2 commented:
> however as provided the first one reads the TRANSLATIONS variable in the qmake project file and processes all the languages to produce what are full translation files with every QString in - for the English (American) language, as the source code is already in that language it produces a much larger file than is needed, and by running the lupdate command with an existing mudlet_en_US.ts file it can be made much simpler with the additional -pluralonly option that only picks out the cases where the %n parameter is used to specify a (usually integer but it need not be) 3rd argument to the QObject::tr() method which enables the handling of all the different plural forms for those many languages where adding an s if there is more (or less) than one article is just plain wrong!

I've read and re-read this _parasentence_ and I'm having difficulty making sense of it. Why do you suggest we exclude singular form of the messages? :thinking:

vadi2 commented:
```
 vadi   Enhance_addGuiTranslations  ⋯  Mudlet  mudlet  src  1  ./unix_lupdate.sh `which lupdate`
Error: /home/vadi/Programs/Qt/5.9/gcc_64/bin/lupdate/ does not appear to be a path to a Qt lupdate executable
 vadi   Enhance_addGuiTranslations  ⋯  Mudlet  mudlet  src  2  which lupdate
/home/vadi/Programs/Qt/5.9/gcc_64/bin/lupdate
```

I'd like to see it just work with the `lupdate` I've got in the `$PATH`. If I want a different binary, I'll just change it like I do for everything else - or at least make it `$PATH` default and have it an option that you can change it.

SlySven commented:
> > _a long and rambling sentance_
>
> I've read and re-read this parasentence and I'm having difficulty making sense of it. Why do you suggest we exclude singular form of the messages?

Briefly: the Qt provide external tool `lupdate` script produces an en_US translation source file (mudlet_en_US.ts) that is is a full-size one with __every__ translatable string included for translation and then needing to be "approved" - even though that **apart from plurals** (and possible "Britishism" coming from me 😊 ) a translation is NOT required.  My script processes that locale specially to make it a _plurals only_ one with something less than twenty texts to approve.  Also, the Qt Script only (_I think_) considers translations included in the `TRANSLATIONS` variable in the qmake project file - my script works through all the files matching the `mudlet_*.ts` globbing pattern.

SlySven commented:
> > Some mojibake which looks as though it is the "usage" output from one of my scripts
>
>I'd like to see it just work with the lupdate I've got in the $PATH. If I want a different binary, I'll just change it like I do for everything else - or at least make it $PATH default and have it an option that you can change it.

I needed to do it like that so that I could ensure with the __multiple__ Qt library versions/kits I have got installed/setup that the right one was used to match the current version selected:
![screenshot_20171119_211607](https://user-images.githubusercontent.com/6163092/32995620-061c72b8-cd6f-11e7-8f95-65427ca57b19.png)

SlySven commented on src/mudlet.cpp:
```diff
@@ -198,142 +283,145 @@ mudlet::mudlet()
+    mpMainToolBar->addAction(mpActionReplay);
+    mpMainToolBar->widgetForAction(mpActionReplay)->setObjectName(mpActionReplay->objectName());
+
+    mpActionDisconnect = new QAction(this);
```
I had to go through that area and clean out a load of cruft relating to unused `QActions` and reorganise what was left - I got the impression that as this appeared on the menu bar and is also present as a lua command that it was something that had been missed out.

I _did_ refrain from including another one! 👼  I wanted to put a "Close" profile button  ![mudlet_door_48px](https://user-images.githubusercontent.com/6163092/32995768-0b6e9df2-cd71-11e7-858d-5c6710fd92b5.png)  that if used returned the application to the initial pre-first profile connection state (with the menu/toolbar buttons disabled - which links into the state things would be at when the changes that #1388 handles):

vadi2 replied:
Not in this PR. Let's have it be a sub-button of `Reconnect` if that's going to be added, we have way too many buttons already as it is.

SlySven commented on src/mudlet.cpp:
```diff
@@ -3250,5 +3720,98 @@ QString mudlet::getMudletPath(const mudletPathType mode, const QString& extra1,
+#if defined( Q_OS_MAC )
+        return QStringLiteral("%1/../Resources/").arg(QCoreApplication::applicationDirPath());
+#else
+        return QStringLiteral("%1/").arg(QCoreApplication::applicationDirPath());
```
Whilst including them in the resource file is one way of doing it and one I am keen to do, I am trying to code my way around something that allows this to be made optional as a resource file method is not amiable to allowing interested parties to create their own language translations and developing them on the fly.

I am still hacking away at this but the implementation as it stands is flawed in that loading up all the translations at start-up is wasteful (keeping all the non-used ones around in memory) and not flexible (changing the source requires restarting the application).  There is a better way of doing it and it includes taking into account the language settings that the user has _already_ told the system to use - but given them the chance to change it if they insist {your tale of changing your resident country and then getting everything in the wrong language suggests something changed automatically - or that you hadn't specified your language settings...! 🙁 }

vadi2 replied:
Well:

> so we don't have to deal with the complexity of find where the translation files are hiding.

Adding both methods is gonna defeat the point here as we'll have even more complicated code to do it _both_ ways! :smile: So nevermind - let's have it be just one way, but at least let's add some search paths for the translations then so we aren't forcing people to make symlinks from whatever folder the Mudlet binary ends up at to the translations.

vadi2 commented:
Alright. What about the Windows and macOS versions of the scripts? Perhaps we won't need them because the web translation service will be automatically pulling the strings anyway so translators don't have to mess around with this complicated stuff.

vadi2 commented:
@SlySven let's work on getting this PR finished out. What is outstanding? The disconnect button has to go as it is offtopic, "none" as a language needs to go, and need to add search paths for loading translations I think is all that is left.

SlySven commented:
I have a revised (by merging in some recent changes) edition that does work for translations included as a (separate) resource file (so the inclusion can be made build environment dependent) which defaults to auto-locale selection (using the user's own or system settings) but which can be changed to a specific one on demand - this is via a comboBox on the first profile preferences ("General") tab - it is then also possible to override the location of the translation's file and specify a manual location (which can be the local source tree) in a `QLineEdit` but which has the same default location (visible as a placeholder text) as the compiled in setting. This seems to work well to me - and I have retained the "none" option but it is right at the end of the options listed (the "auto", default locale/language one is right at the start).

However given the textual changes elsewhere and the other things like module code that this current includes but which I developed further/better in #1355 - I would like to get that one in and then redo this one with less in it based upon the current development branch.  Basically I think in merging in things I may have broken other PRs and want to try again now I have something that operates as I would like... hopefully for Mudlet 3.7 .  I have had to juggle some things on the profile preferences (added another tab for "Profile and Server preferences") and with all the other changes that have happened there it will need careful going over again to get this in:

(For testing purposes I set up another user on my Box _stephane_ who has a locale setting of "fr;en_GB") on there the current code does select things to be French - once I remember to have the translations where they should be:

![screenshot_20171204_172921](https://user-images.githubusercontent.com/6163092/33567521-63c4f0f4-d91b-11e7-87f1-f377274e5295.png)

There is a bit of a mix up that I have to check - but the default location for this build should have been in `:/translations` which is a resource file within the application, which hitting the clear button `⌫` should have shown:
![screenshot_20171204_173054](https://user-images.githubusercontent.com/6163092/33567520-637d0ea6-d91b-11e7-85eb-acc2639bfdab.png)

This has to use a non-native file dialog so that the _directory_ search will show only the `mudlet_*.qm` files that are wanted:

![screenshot_20171204_173456](https://user-images.githubusercontent.com/6163092/33567519-631fb99a-d91b-11e7-8bb1-661b7ef533fb.png)

vadi2 commented:
OK, thank you, great progress. I've got feedback on this but as you suggest let's process https://github.com/Mudlet/Mudlet/pull/1355 first.

SlySven commented:
OK, once #1355 and #1446 finally get in I can get back to bringing this Frankensteinian monster to life again.  Those other two have things that interact with this one (I touched the package/module handling stuff in this one but it was only completed in the first one so that need merging in correctly and the font optional font handling stuff which has been done by moving the fonts to a second, optional, resource file has shown me how to place the translations into the resource system - for _some_ platforms and how to access them afterwards).

🎉 All in all, I have good hopes for progressing Mudlet in 2018...! 🎆

SlySven commented:
Aaaarrggh! :hurtrealbad: As a test I have tried to merge the development branch into this on my local working code - and, I _cannot_ do it - there are just to many other changes to keep track of.  I am going to try and go through the changes racked up here bit-by- bit and cherry-pick them onto a fresh copy of _development_ with, I think, 1355 and 1446 merged in first...
